src/rate_limiter.rs

```diff
@@ -274,114 +557,116 @@ mod tests {
-
-        // Advance the clock by just less than the same amount again and check that neither client
-        // still cannot add a large request.
-        FakeClock::advance_time(wait_millis - 1);
-        assert_large_message_cannot_be_added(&mut rate_limiter, &client1);
-        assert_large_message_cannot_be_added(&mut rate_limiter, &client2);
-
-        // Advance the clock a final small amount and check that both clients can now add large
-        // requests.
-        FakeClock::advance_time(2);
-        assert_large_message_can_be_added(&mut rate_limiter, &client1);
-        assert_large_message_can_be_added(&mut rate_limiter, &client2);
+        assert!(sized_message_can_be_added(
+            SOFT_CAPACITY / 2,
+            &mut rate_limiter,
+            &client2,
+        ));
+        assert!(sized_message_cannot_be_added(
+            SOFT_CAPACITY / 2,
+            &mut rate_limiter,
+            &client1,
+        ));
+        assert!(small_message_can_be_added(&mut rate_limiter, &client1));
     }
 
+    // FIXME(michael): this is slightly broken
     /// Checks that if two clients add messages with a delay between them, the rate-limiter's
     /// throughput remains constant, but the per-client throughput drops when both clients have
     /// messages and increases when just one has messages.
     #[test]
     fn staggered_start() {
         let mut rate_limiter = RateLimiter::new(false);
-        let client1 = IpAddr::from([0, 0, 0, 0]);
-        let client2 = IpAddr::from([1, 1, 1, 1]);
 
-        // This is the time during which half of a large request will leak away. Assert this
-        // is the case by adding a large request, then waiting for two times `wait_millis` then
-        // trying to add a large request again.
-        let wait_millis = (CLIENT_CAPACITY * 500 / RATE as u64) + 1; // round up
-        assert_large_message_can_be_added(&mut rate_limiter, &client1);
-        FakeClock::advance_time(wait_millis);
-        assert_large_message_cannot_be_added(&mut rate_limiter, &client1);
-        FakeClock::advance_time(wait_millis);
-        assert_large_message_can_be_added(&mut rate_limiter, &client1);
+        // Saturate the rate limiter so that every client's cap is reduced to MIN_CLIENT_CAPACITY.
+        let num_clients = (SOFT_CAPACITY as f64 / MIN_CLIENT_CAPACITY as f64).ceil() as u64;
 
-        // Allow the rate-limiter to empty.
-        FakeClock::advance_time(2 * wait_millis);
+        println!("num_clients = {}", num_clients);
+        let clients: Vec<_> = (0..num_clients as u8)
+            .map(|i| IpAddr::from([i, i, i, i]))
+            .collect();
 
-        // Client 1 adds a large message then after `wait_millis`, Client 2 does likewise.
-        assert_large_message_can_be_added(&mut rate_limiter, &client1);
-        FakeClock::advance_time(wait_millis);
-        assert_large_message_can_be_added(&mut rate_limiter, &client2);
+        // All clients put a MIN_CLIENT_CAPACITY-sized message.
```

Comment:
`MIN_CLIENT_CAPACITY sized` ?  also I think MIN_CLIENT_CAPACITY shall be surrounded by tick?